rfcs/20230321-fp8_fnuz.md

```diff
@@ -0,0 +1,165 @@
+# RFC: Float8E4M3FNUZ and Float8E5M2FNUZ
+
+## Summary
+
+Graphcore, AMD, and Qualcomm have proposed two new FP8 types, Float8E4M3FNUZ
+and Float8E5M2FNUZ[^1]. These types are implemented in commercially available
+hardware[^2], and added to MLIR builtin types[^4] and LLVM APFloat[^5].
+
+These two types appear similar to the existing types Float8E4M3FN and
+Float8E5M2[^3], but differ in important ways.
+
+## Details
+
+Both Float8E4M3FNUZ and Float8E5M2FNUZ differ from typical floating point types
+in their support for NaN, infinities, negative zero, and exponent bias. The
+suffix "FNUZ" is derived from these differences. `F` is for "finite" (no
+infinities), `N` for with special NaN encoding, `UZ` for unsigned zero.
```

Comment:
Can you be a bit more specific here about the bias?  i.e. that it is one higher than typical IEEE-754?  Is this part of the `UZ` naming convention - as in all types labelled with `UZ` would have this assumption?  Otherwise I would have expected the non-standard bias to also be reflected in the name.

Reply:
Thanks for taking a look. I'll clear it up in the RFC, but also explain here.

The change to the bias is a result of having an additional exponent encoding freed up by not having `inf` and only one `NaN` encoding. We decide to use this "extra exponent" at the bottom of the range, so we bias by one extra value.

Because of the "extra exponent" we have the same max value of a IEEE-754-like FP8 type. This can be seen in the comparison of `Float8E5M2` and `Float8E5M2FNUZ`. They both have maximum value 57344, but `Float8E5M2FNUZ` can go down to 0.25 $\times$ 2<sup>-15</sup> = 2<sup>-17</sup>.

The use of the additional encoding space is a free design choice.

Reply:
Perhaps another way to say it is these proposed types have the same absolute maximum normal number values as a hypothetical IEEE FP8 type with the same number of exponent bits, with extra precision near zero. Given that users interact with floating point numbers as decimal values and not their binary encodings, that seems more intuitively aligned with the IEEE types.

Reply:
Yes, I understand the reasoning - and maybe this discussion was already had for the MLIR/LLVM types - if it was a free choice, I would have expected it to be reflected in the name - just as you describe for the other differences F,N,UZ.

My question was more about the choice of name.  I do think we should keep the current naming consistent with MLIR/LLVM, so I'm not proposing to change it now. 

My request for the RFC though it just to explicitly call out the exact difference in bias here (i.e. increased by one), and maybe add a short comment similar to above, like
> ... in order to keep a similar dynamic range

Reply:
I've update the doc. Let me know if you prefer a different wording.

Reply:
sg, this is exactly what I meant, thanks.